doc/developer-guides/sw_design_guidelines.rst

```diff
@@ -0,0 +1,494 @@
+   |                    |                         |              |                           | such as IO request      |
+   |                    |                         |              |                           | buffers and sbuf for    |
+   |                    |                         |              |                           | debug.                  |
+   +--------------------+-------------------------+--------------+---------------------------+-------------------------+
+   | External resource  | Invalid E820 table or   | Yes          | The hypervisor shall      | Invalid E820 table or   |
+   | provided by        | invalid boot information|              | panic during platform     | invalid boot information|
+   | bootloader         |                         |              | initialization.           | .                       |
+   | (UEFI or SBL)      |                         |              |                           |                         |
+   +--------------------+-------------------------+--------------+---------------------------+-------------------------+
+   | Physical resource  | 1GB page is not         | Yes          | The hypervisor shall      | 1GB page is not         |
+   | used by the        | available on the        |              | panic during platform     | available on the        |
+   | hypervisor         | platform or invalid     |              | initialization.           | platform or invalid     |
+   |                    | physical CPU ID         |              |                           | physical CPU ID.        |
+   +--------------------+-------------------------+--------------+---------------------------+-------------------------+
+
+
+Examples
+--------
+
+Here is an example to illustrate when error handling codes are required on
+architecture level.
+
+There are two pre-condition statements of ``vcpu_from_vid``. It indicates that
+it's the caller's responsibility to guarantee these pre-conditions.
+
+.. code-block:: c
+
+  /*
+   * @pre vcpu_id < CONFIG_MAX_VCPUS_PER_VM
+   * @pre &(vm->hw.vcpu_array[vcpu_id])->state != VCPU_OFFLINE
+   */
+  static inline struct acrn_vcpu *vcpu_from_vid(struct acrn_vm *vm, uint16_t vcpu_id)
+  {
+          return &(vm->hw.vcpu_array[vcpu_id]);
+  }
+
+``vcpu_from_vid`` is called by ``hcall_set_vcpu_regs``, which is a hypercall.
+``hcall_set_vcpu_regs`` is an external interface and ``vcpu_id`` is provided by
+VM. In this case, we shall add the error checking codes before calling
+``vcpu_from_vid`` to make sure that the passed parameters are valid and the
+pre-conditions are guaranteed.
+
+Here is the sample codes for error checking before calling ``vcpu_from_vid``:
+
+.. code-block:: c
+
+  if (vcpu_id >= CONFIG_MAX_VCPUS_PER_VM) {
+          pr_err("vcpu id is out of range \r\n");
+          status = -EINVAL;
+  }
+
+  if ((&(vm->hw.vcpu_array[vcpu_id]))->state == VCPU_OFFLINE) {
```

Comment:
As written, if ``vcpu_id`` is invalid (as detected by the first test), this code will still be run with an invalid value for ``vcpu_id``.   This should be in an else if.